internal/controller/etcdcluster_controller.go

```diff
@@ -36,24 +44,349 @@ type EtcdClusterReconciler struct {
 //+kubebuilder:rbac:groups=etcd.aenix.io,resources=etcdclusters,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=etcd.aenix.io,resources=etcdclusters/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=etcd.aenix.io,resources=etcdclusters/finalizers,verbs=update
+//+kubebuilder:rbac:groups="",resources=configmaps,verbs=get;list;watch;create;update;watch;delete;patch
+//+kubebuilder:rbac:groups="",resources=services,verbs=get;create;delete;update;patch;list;watch
+//+kubebuilder:rbac:groups="apps",resources=statefulsets,verbs=get;create;delete;update;patch;list;watch
 
-// Reconcile is part of the main kubernetes reconciliation loop which aims to
-// move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the EtcdCluster object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.0/pkg/reconcile
+// Reconcile checks CR and current cluster state and performs actions to transform current state to desired.
 func (r *EtcdClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx)
+	logger.V(2).Info("reconciling object", "namespaced_name", req.NamespacedName)
+	instance := &etcdaenixiov1alpha1.EtcdCluster{}
+	err := r.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.V(2).Info("object not found", "namespaced_name", req.NamespacedName)
+			return ctrl.Result{}, nil
+		}
+		// Error retrieving object, requeue
+		return reconcile.Result{}, err
+	}
+	// If object is being deleted, skipping reconciliation
+	if instance.DeletionTimestamp != nil && !instance.DeletionTimestamp.IsZero() {
```

Comment:
Comparison with nil is redundant, due to `IsZero()` method already includes this check.